source/common/upstream/cluster_manager_impl.cc

```diff
@@ -238,38 +263,13 @@ ClusterManagerImpl::ClusterManagerImpl(const envoy::config::bootstrap::v2::Boots
     }
   }
 
-  Optional<std::string> local_cluster_name;
-  if (!cm_config.local_cluster_name().empty()) {
-    local_cluster_name_ = cm_config.local_cluster_name();
-    local_cluster_name.value(cm_config.local_cluster_name());
-    if (primary_clusters_.find(local_cluster_name.value()) == primary_clusters_.end()) {
-      throw EnvoyException(
-          fmt::format("local cluster '{}' must be defined", local_cluster_name.value()));
-    }
-  }
-
   // Once the initial set of static bootstrap clusters are created (including the local cluster),
   // we can instantiate the thread local cluster manager.
   tls_->set([this, local_cluster_name](
                 Event::Dispatcher& dispatcher) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return std::make_shared<ThreadLocalClusterManagerImpl>(*this, dispatcher, local_cluster_name);
   });
 
-  // Now setup ADS if needed, this might rely on a primary cluster and the
-  // thread local cluster manager.
```

Comment:
This comment was there to point out that ADS init must happen after the above `tls_->set`. We need TLS to be able to initialize gRPC async client manager bits. I'm surprised the tests pass without this (specifically `ads_integration_test` with Google gRPC).

Reply:
Actually, I think the need for this went away when I fixed the stats scope to be created on the main thread. Feel free to ignore this comment, if an issue arises I can go and fix it, since this is `Grpc::AsyncClient` implementation stuff.